test/endpoint_testrunner.py

```diff
@@ -0,0 +1,238 @@
+    """
+    Return which Inference GPU canary images to run canary tests on for a given framework and AWS region
+
+    :param framework: ML framework (mxnet, tensorflow, pytorch)
+    :param region: AWS region
+    :param image_type: inference
+    :return: a list of space separated image uris
+    """
+    customer_type_tag = "-sagemaker"
+    version_regex = {
+        "tensorflow": rf"tf(-sagemaker)?{customer_type_tag}-(\d+.\d+)",
+        "mxnet": rf"mx(-sagemaker)?{customer_type_tag}-(\d+.\d+)",
+        "pytorch": rf"pt(-sagemaker)?{customer_type_tag}-(\d+.\d+)",
+    }
+
+    # Get tags from repo releases
+    repo = git.Repo(os.getcwd(), search_parent_directories=True)
+
+    versions_counter = {}
+    pre_populated_py_version = {}
+
+    for tag in repo.tags:
+        tag_str = str(tag)
+        
+        match = re.search(version_regex[framework], tag_str)
+        ## The tags not having -py3 will not pass the condition below
+        ## This eliminates all the old and testing tags that we are not monitoring.
+        if match:
+            version = match.group(2)
+            if not versions_counter.get(version):
+                versions_counter[version] = {"tr": False, "inf": False}
+
+            if "tr" not in tag_str and "inf" not in tag_str:
+                versions_counter[version]["tr"] = True
+                versions_counter[version]["inf"] = True
+            elif "tr" in tag_str:
+                versions_counter[version]["tr"] = True
+            elif "inf" in tag_str:
+                versions_counter[version]["inf"] = True
+
+            try:
+                python_version_extracted_through_regex = match.group(3)
+                if python_version_extracted_through_regex:
+                    if version not in pre_populated_py_version:
+                        pre_populated_py_version[version] = set()
+                    pre_populated_py_version[version].add(python_version_extracted_through_regex)
+            except IndexError:
+                LOGGER.info(f"For Framework: {framework} we do not use regex to fetch python version")
+
+    versions = []
+    for version, inf_train in versions_counter.items():
+        if inf_train["inf"]:
+            versions.append(version)
+
+    # Sort ascending to descending, use lambda to ensure 2.2 < 2.15, for instance
+    versions.sort(key=lambda version_str: [int(point) for point in version_str.split(".")], reverse=True)
+
+    framework_versions = versions if len(versions) < 4 else versions[:3]
+    image_definitions = []
+    for fw_version in framework_versions:
+        if fw_version in pre_populated_py_version:
+            py_versions = pre_populated_py_version[fw_version]
+        else:
+            py_versions = [get_canary_default_tag_py3_version(framework, fw_version)]
+        for py_version in py_versions:
+            image_definition = []
+            if processor == "gpu" or processor is None:
```

Comment:
1. How is the action different between the cases where `processor == "gpu"` and `processor == "cpu"`? The `dict` object added to the `image_definitions` list seems to be identical in both cases.
2. If `processor is None`, then the code will always pass through both the first if-condition as well as the second if-condition, adding 2 identical entries into the `image_definitions` list, both with `processor` set to `None`. How is this useful?

Reply:
Good catch! The logic was designed to capture both gpu and cpu images only when processor is None. The bug has been fixed.